default/python/AI/PolicyAI.py

```diff
@@ -0,0 +1,242 @@
+
+    def _techno_or_industry(self):
+        research_prio = self._aistate.get_priority(PriorityType.RESOURCE_RESEARCH)
+        production_prio = self._aistate.get_priority(PriorityType.RESOURCE_PRODUCTION)
+        may_switch_to_techno = self._can_adopt(technocracy, industrialism)
+        may_switch_to_industry = self._can_adopt(industrialism, technocracy)
+        # do not switch too often
+        if technocracy in self._adopted:
+            research_prio *= 1.25
+            may_switch_to_techno = False
+        elif industrialism in self._adopted:
+            production_prio *= 1.25
+            may_switch_to_industry = False
+        if may_switch_to_industry and research_prio < production_prio:
+            self._deadopt(technocracy)
+            self._adopt(industrialism)
+        elif may_switch_to_techno and research_prio > production_prio:
+            self._deadopt(industrialism)
+            self._adopt(technocracy)
+        debug(
+            "Prio res./ind. = %.1f/%.1f, maySwitch = %s/%s",
+            research_prio,
+            production_prio,
+            may_switch_to_techno,
+            may_switch_to_industry,
+        )
+
+    def _determine_influence_priority(self, new_production):
+        # priority shouldn't only be based on current value
+        forecast = self._ip + 3 * new_production
+        if infra1 in self._adopted and bureaucracy not in self._adopted and centralization not in self._adopted:
+            forecast -= self._num_populated + self._centralization_cost + self._bureaucracy_cost
+        if forecast < 40:
+            prio = 60 - forecast
+        else:
+            prio = 800 / forecast
+        prio *= (3 + self._num_populated + self._num_outposts / 10) / 10
+        # if not infra3 in self._adopted: prio += 10
+        debug("Setting influcence priority to %.1f", prio)
+        self._aistate.set_priority(PriorityType.RESOURCE_INFLUENCE, prio)
+
+    def _get_infl_prod(self, update=False):
+        """Get / Update IP production."""
+        if update:
+            self._universe.updateMeterEstimates(self._populated_planet_ids)
+            fo.updateResourcePools()
+        return self._empire.resourceProduction(fo.resourceType.influence)
+
+    def _can_adopt(self, name, replace_other=None):
+        policy = fo.getPolicy(name)
+        ret = name in self._empire.availablePolicies and policy.adoptionCost() <= self._ip
+        if replace_other and replace_other in self._adopted:
+            return ret
+        return (
+            ret and self._empire.emptyPolicySlots[policy.category] and self._empire.policyPrereqsAndExclusionsOK(name)
+        )
+
+    def _get_adoptable(self):
+        """List of adoptable policies with the still available IP."""
+        ret = set()
+        for p in self._empire.availablePolicies:
+            if p in self._adopted:
+                continue
+            if self._can_adopt(p):
+                ret.add(p)
+        return ret
```

Comment:
Set comprehension is a great thing.
```suggestion
        def can_adopt(policy):
            return policy not in self._adopted and self._can_adopt(policy)

        return {p for p in self._empire.availablePolicies if can_adopt(p)}
```

Reply:
Actually I think _can_adopt should return False, for a policy that is already adopted, that would simplify other functions, too.

I don't know why I didn't think about comprehension before (well, apart from the fact that I've never heard the name, but I have used it). Will change.